### DISCUSSION TITLE
Vector<T> and AVX512

### DISCUSSION THREAD

**Discussion by gfoidl** [dotnet/runtime#85155] [Q&A] [answered]
Currently `Vector<T>` has the size of the largest available (vector) register (so either 128 bit or 256 bit).
With AVX512 coming to .NET will `Vector<T>` "grow" with it?

@tannergooding can you please shed some light into this (even if it's too early to manifest it, but I'd like to see where the journey goes)?

**Comment by tannergooding** [comment, marked as answer]
`Vector<T>` will likely not support implicit growth to `Vector512<T>`. We will look into and may provide some explicit mechanism through which to opt into the functionality.

There are two primary considerations at play here.

First, a minor concern is that some developers may have written code which presumes no larger than 32-byte `Vector<T>`. Thus, growing the size of `Vector<T>` may break their existing code. This is a bug in their code, but it is a concern we must factor in nonetheless.

Second, implicit growth to 64-byte `Vector<T>` may negatively impact the performance of some code. This is particularly the case for inputs between `1-63` bytes (or sometimes `1-127` bytes, as `Count * 2` is a common check) which may no longer be vectorized where they were previously. If such "small" inputs were regular, such as for processing names, phone numbers, etc; then this may silently be a perf reduction. This can also impact "trailing" data that does not fit into a `Vector<T>` and thus may fallback to a scalar loop if the developer did not explicitly go the extra mile to handle it as an idempotent backtracked operation (and thus in a single additional vectorized iteration). Finally it can also be impacted due to the heavy downclocking seen on the initial hardware that introduced 512-bit support. This is of course almost non-existent on the newer hardware, but remains a general concern for some subset of hardware that may still be used in varying contexts.

**Reply by gfoidl** [reply]
Perfect, thanks 👍🏻
